daemon/policy.go

```diff
@@ -237,28 +235,20 @@ func (d *Daemon) PolicyAdd(rules api.Rules, opts *AddOptions) (uint64, error) {
 	prefixes := policy.GetCIDRPrefixes(rules)
 	log.WithField("prefixes", prefixes).Debug("Policy imported via API, found CIDR prefixes...")
 
-	prefixIdentities, err := identity.AllocateCIDRIdentities(prefixes)
-	if err != nil {
-		metrics.PolicyImportErrors.Inc()
-		return d.policy.GetRevision(), err
-	}
-
-	if err = ipcache.CheckPrefixes(bpfIPCache.IPCache, prefixes); err == nil {
-		err = ipcache.UpsertIPNetsToKVStore(prefixes, prefixIdentities)
-	}
-	if err != nil {
+	if err := ipcache.AllocateCIDRs(bpfIPCache.IPCache, prefixes); err != nil {
 		metrics.PolicyImportErrors.Inc()
-
-		// Failed to update Prefix->ID mappings; don't leak identities
-		identity.ReleaseSlice(prefixIdentities)
+		log.WithError(err).WithField("prefixes", prefixes).Warning(
+			"Failed to allocate identities for CIDRs during policy add")
 		return d.policy.GetRevision(), err
 	}
 
 	rev, err := d.policyAdd(rules, opts)
 	if err != nil {
 		// Don't leak identities allocated above.
-		identity.ReleaseSlice(prefixIdentities)
-
+		if err2 := ipcache.ReleaseCIDRs(prefixes); err2 != nil {
```

Comment:
This was the earlier behaviour, I think, but if we already had some CIDRs added, and we see a failure for a new set of CIDRs that include those older CIDRs, this would then remove those as well, right?

Reply:
Also, doesn't AllocateCIDRs do the release for us on error?

Reply:
Oh, I guess there is some local/remote difference that is too complicated for me to understand.

Reply:
Hmm. In this case all of the CIDRs for this policy were successfully added, but then during import of the policy into the repository it was rejected so this should now roll back all of the kvstore interactions - both for the identitie allocations (which was previously done) and for the ipcache mappings (which was previously skipped).

We're using the same slice of prefixes here as were used during allocation, so I'd expect this to release the identities and mappings that were just created in the kvstore.

With regards to interactions with prefixes other CIDR policies, the theory is that each CIDR rule takes its own reference on each kvstore key and the delete only occurs in the kvstore if all of them are deleted. That said I suspect there's another bug in that area with how the ipcache is managed that means this is not quite the case. I plan to follow up on that separately, as I think there are ways to trigger via other means - that is, addition of two CIDR rules followed by removal of one.

Reply:
Here's the PR: #4611